File: python/postfiat/envelope/storage.py
"""Storage backends for PostFiat envelope content.

Provides different strategies for storing and retrieving content.
"""

from abc import ABC, abstractmethod
from typing import Tuple, List
import hashlib
import uuid
import json
import base64
from ..v3.messages_pb2 import (
    ContentDescriptor, Envelope, CoreMessage, 
    MultiPartMessagePart, MessageType, EncryptionMode
)
from ..exceptions import ValidationError


class ContentStorage(ABC):
    """Abstract base class for content storage strategies."""
    
    @abstractmethod
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Store content and return a descriptor.
        
        Args:
            content: Raw content bytes to store
            content_type: MIME type of the content
            
        Returns:
            ContentDescriptor with URI and metadata
        """
        pass
    
    @abstractmethod
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve content from storage.
        
        Args:
            descriptor: Content descriptor with URI
            
        Returns:
            Raw content bytes
        """
        pass
    
    @abstractmethod
    def can_handle(self, uri: str) -> bool:
        """Check if this storage can handle the given URI scheme.
        
        Args:
            uri: URI to check
            
        Returns:
            True if this storage can handle the URI
        """
        pass


class IPFSStorage(ContentStorage):
    """IPFS content storage implementation."""
    
    def __init__(self, gateway_url: str = "http://localhost:5001"):
        """Initialize IPFS storage.
        
        Args:
            gateway_url: IPFS API gateway URL
        """
        self.gateway_url = gateway_url
        self._client = None
    
    @property
    def client(self):
        """Lazy-load IPFS client."""
        if self._client is None:
            try:
                import ipfshttpclient
                self._client = ipfshttpclient.connect(self.gateway_url)
            except ImportError:
                raise ImportError(
                    "ipfshttpclient is required for IPFS storage. "
                    "Install with: pip install postfiat-sdk[storage]"
                )
            except Exception as e:
                raise ConnectionError(f"Failed to connect to IPFS at {self.gateway_url}: {e}")
        return self._client
    
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Store content in IPFS."""
        # Calculate content hash
        content_hash = hashlib.sha256(content).digest()
        
        try:
            # Add content to IPFS
            result = self.client.add_bytes(content)
            cid = result
            
            return ContentDescriptor(
                uri=f"ipfs://{cid}",
                content_type=content_type,
                content_length=len(content),
                content_hash=content_hash,
                metadata={
                    "storage_provider": "ipfs",
                    "gateway_url": self.gateway_url
                }
            )
        except Exception as e:
            # Fallback to simulated CID if IPFS is not available
            cid = f"Qm{content_hash.hex()[:44]}"
            return ContentDescriptor(
                uri=f"ipfs://{cid}",
                content_type=content_type,
                content_length=len(content),
                content_hash=content_hash,
                metadata={
                    "storage_provider": "ipfs",
                    "gateway_url": self.gateway_url,
                    "simulated": "true",
                    "error": str(e)
                }
            )
    
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve content from IPFS."""
        if not self.can_handle(descriptor.uri):
            raise ValidationError(f"Invalid IPFS URI: {descriptor.uri}")
        
        # Extract CID from URI
        cid = descriptor.uri[7:]  # Remove "ipfs://" prefix
        
        try:
            # Retrieve content from IPFS
            return self.client.cat(cid)
        except Exception as e:
            raise IOError(f"Failed to retrieve content from IPFS: {e}")
    
    def can_handle(self, uri: str) -> bool:
        """Check if URI is an IPFS URI."""
        return uri.startswith("ipfs://")


class MultipartStorage(ContentStorage):
    """Storage-agnostic multipart message chunking.
    
    This storage backend has no knowledge of the underlying ledger.
    It just creates multipart envelopes that can be stored anywhere.
    """
    
    def __init__(self, max_part_size: int = 800):
        """Initialize multipart storage.
        
        Args:
            max_part_size: Maximum size per envelope part in bytes
        """
        self.max_part_size = max_part_size
    
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Create multipart descriptor - no actual storage happens here."""
        # Generate unique message ID
        message_id = str(uuid.uuid4())
        
        # Calculate content hash
        content_hash = hashlib.sha256(content).digest()
        
        # Calculate number of parts needed
        total_parts = (len(content) + self.max_part_size - 1) // self.max_part_size
        
        # Create multipart URI - describes how to find all parts
        uri = f"multipart://{message_id}"
        
        return ContentDescriptor(
            uri=uri,
            content_type=content_type,
            content_length=len(content),
            content_hash=content_hash,
            metadata={
                "storage_provider": "multipart",
                "message_id": message_id,
                "total_parts": str(total_parts),
                "part_size": str(self.max_part_size)
            }
        )
    
    def create_part_envelopes(
        self, 
        content: bytes, 
        descriptor: ContentDescriptor,
        encryption_mode: EncryptionMode = EncryptionMode.PROTECTED,
        base_metadata: dict = None
    ) -> List[Envelope]:
        """Create multipart envelopes for ledger storage.
        
        Args:
            content: Content to split into parts
            descriptor: Content descriptor with multipart metadata
            encryption_mode: Encryption mode for envelopes
            base_metadata: Base metadata for all envelopes
            
        Returns:
            List of envelopes containing multipart message parts
        """
        if not self.can_handle(descriptor.uri):
            raise ValidationError(f"Invalid ledger URI: {descriptor.uri}")
        
        message_id = descriptor.metadata.get("message_id")
        total_parts = int(descriptor.metadata.get("total_parts", "0"))
        
        if not message_id or total_parts == 0:
            raise ValidationError("Missing multipart metadata")
        
        envelopes = []
        
        for i in range(0, len(content), self.max_part_size):
            part_number = (i // self.max_part_size) + 1
            chunk_content = content[i:i + self.max_part_size]
            
            # Create multipart message part
            multipart = MultiPartMessagePart(
                message_id=message_id,
                part_number=part_number,
                total_parts=total_parts,
                content=chunk_content,
                complete_message_hash=descriptor.content_hash.hex()
            )
            
            # Create envelope for this part
            message_bytes = multipart.SerializeToString()
            
            metadata = (base_metadata or {}).copy()
            metadata.update({
                "multipart": f"{part_number}/{total_parts}",
                "message_id": message_id
            })
            
            envelope = Envelope(
                version=1,
                content_hash=hashlib.sha256(message_bytes).digest(),
                message_type=MessageType.MULTIPART_MESSAGE_PART,
                encryption=encryption_mode,
                message=message_bytes,
                metadata=metadata
            )
            
            envelopes.append(envelope)
        
        return envelopes
    
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve requires collecting all parts - not implemented here."""
        raise NotImplementedError(
            "Multipart retrieval requires collecting all envelope parts by message_id"
        )
    
    def can_handle(self, uri: str) -> bool:
        """Check if URI is a multipart URI."""
        return uri.startswith("multipart://")


class HTTPStorage(ContentStorage):
    """HTTP/HTTPS content storage implementation."""
    
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Store content via HTTP POST."""
        # This would typically POST to a content server
        raise NotImplementedError("HTTP storage not yet implemented")
    
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve content via HTTP GET."""
        if not self.can_handle(descriptor.uri):
            raise ValidationError(f"Invalid HTTP URI: {descriptor.uri}")
        
        # TODO: Implement HTTP retrieval
        # import requests
        # response = requests.get(descriptor.uri)
        # return response.content
        
        raise NotImplementedError("HTTP retrieval not yet implemented")
    
    def can_handle(self, uri: str) -> bool:
        """Check if URI is an HTTP/HTTPS URI."""
        return uri.startswith("http://") or uri.startswith("https://")


class CompositeStorage(ContentStorage):
    """Composite storage that delegates to appropriate backend."""
    
    def __init__(self, storages: List[ContentStorage]):
        """Initialize with list of storage backends.
        
        Args:
            storages: List of storage backends to use
        """
        self.storages = storages
    
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Store using the first available storage backend."""
        if not self.storages:
            raise ValidationError("No storage backends configured")
        
        # Use first storage backend for storing
        return self.storages[0].store(content, content_type)
    
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve using appropriate storage backend based on URI."""
        for storage in self.storages:
            if storage.can_handle(descriptor.uri):
                return storage.retrieve(descriptor)
        
        raise ValidationError(f"No storage backend can handle URI: {descriptor.uri}")
    
    def can_handle(self, uri: str) -> bool:
        """Check if any storage backend can handle the URI."""
        return any(storage.can_handle(uri) for storage in self.storages)


class RedisStorage(ContentStorage):
    """Redis content storage implementation."""
    
    def __init__(self, redis_url: str = "redis://localhost:6379", key_prefix: str = "postfiat:content"):
        """Initialize Redis storage.
        
        Args:
            redis_url: Redis connection URL
            key_prefix: Prefix for Redis keys
        """
        self.redis_url = redis_url
        self.key_prefix = key_prefix
        self._client = None
    
    @property
    def client(self):
        """Lazy-load Redis client."""
        if self._client is None:
            try:
                import redis
                self._client = redis.from_url(self.redis_url)
            except ImportError:
                raise ImportError(
                    "redis is required for Redis storage. "
                    "Install with: pip install postfiat-sdk[storage]"
                )
        return self._client
    
    def _content_key(self, content_hash: bytes) -> str:
        """Get Redis key for content."""
        return f"{self.key_prefix}:{content_hash.hex()}"
    
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Store content in Redis."""
        # Calculate content hash
        content_hash = hashlib.sha256(content).digest()
        
        # Store content in Redis
        key = self._content_key(content_hash)
        
        # Store content and metadata together
        content_data = {
            "content": base64.b64encode(content).decode(),
            "content_type": content_type,
            "content_length": len(content),
            "content_hash": content_hash.hex()
        }
        
        try:
            self.client.set(key, json.dumps(content_data))
            
            return ContentDescriptor(
                uri=f"redis://{content_hash.hex()}",
                content_type=content_type,
                content_length=len(content),
                content_hash=content_hash,
                metadata={
                    "storage_provider": "redis",
                    "redis_key": key,
                    "redis_url": self.redis_url
                }
            )
        except Exception as e:
            raise IOError(f"Failed to store content in Redis: {e}")
    
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve content from Redis."""
        if not self.can_handle(descriptor.uri):
            raise ValidationError(f"Invalid Redis URI: {descriptor.uri}")
        
        # Extract content hash from URI
        content_hash_hex = descriptor.uri[8:]  # Remove "redis://" prefix
        content_hash = bytes.fromhex(content_hash_hex)
        
        key = self._content_key(content_hash)
        
        try:
            content_data = self.client.get(key)
            if content_data is None:
                raise IOError(f"Content not found in Redis: {descriptor.uri}")
            
            data = json.loads(content_data)
            content = base64.b64decode(data["content"])
            
            # Verify content hash
            if hashlib.sha256(content).digest() != content_hash:
                raise IOError(f"Content hash mismatch for {descriptor.uri}")
            
            return content
        except Exception as e:
            raise IOError(f"Failed to retrieve content from Redis: {e}")
    
    def can_handle(self, uri: str) -> bool:
        """Check if URI is a Redis URI."""
        return uri.startswith("redis://")


class InlineStorage(ContentStorage):
    """Inline content storage - stores content directly in descriptor."""
    
    def store(self, content: bytes, content_type: str) -> ContentDescriptor:
        """Store content inline in descriptor."""
        # Calculate content hash
        content_hash = hashlib.sha256(content).digest()
        
        # Encode content as base64 for inline storage
        content_b64 = base64.b64encode(content).decode()
        
        return ContentDescriptor(
            uri=f"inline://data",
            content_type=content_type,
            content_length=len(content),
            content_hash=content_hash,
            metadata={
                "storage_provider": "inline",
                "content_data": content_b64
            }
        )
    
    def retrieve(self, descriptor: ContentDescriptor) -> bytes:
        """Retrieve content from descriptor metadata."""
        if not self.can_handle(descriptor.uri):
            raise ValidationError(f"Invalid inline URI: {descriptor.uri}")
        
        try:
            content_b64 = descriptor.metadata.get("content_data")
            if not content_b64:
                raise IOError(f"No inline content data found in descriptor")
            
            content = base64.b64decode(content_b64)
            
            # Verify content hash
            if hashlib.sha256(content).digest() != descriptor.content_hash:
                raise IOError(f"Content hash mismatch for inline content")
            
            return content
        except Exception as e:
            raise IOError(f"Failed to retrieve inline content: {e}")
    
    def can_handle(self, uri: str) -> bool:
        """Check if URI is an inline URI."""
        return uri.startswith("inline://")

File: typescript/src/envelope/storage.ts
/**
 * Storage backends for PostFiat envelope content.
 * 
 * Provides different strategies for storing and retrieving content.
 */

import { ContentDescriptor, Envelope, CoreMessage, MultiPartMessagePart, MessageType, EncryptionMode } from '../generated/postfiat/v3/messages_pb';
import { createHash } from 'crypto';

export class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

/**
 * Abstract base class for content storage strategies.
 */
export abstract class ContentStorage {
  /**
   * Store content and return a descriptor.
   */
  abstract store(content: Uint8Array, contentType: string): Promise<ContentDescriptor>;

  /**
   * Retrieve content from storage.
   */
  abstract retrieve(descriptor: ContentDescriptor): Promise<Uint8Array>;

  /**
   * Check if this storage can handle the given URI scheme.
   */
  abstract canHandle(uri: string): boolean;
}

/**
 * IPFS content storage implementation.
 */
export class IPFSStorage extends ContentStorage {
  private gatewayUrl: string;
  private _client: any;

  constructor(gatewayUrl: string = 'http://localhost:5001') {
    super();
    this.gatewayUrl = gatewayUrl;
  }

  private async getClient() {
    if (!this._client) {
      try {
        const { create } = await import('ipfs-http-client');
        this._client = create({ url: this.gatewayUrl });
      } catch (error) {
        throw new Error(
          'ipfs-http-client is required for IPFS storage. ' +
          'Install with: npm install ipfs-http-client'
        );
      }
    }
    return this._client;
  }

  async store(content: Uint8Array, contentType: string): Promise<ContentDescriptor> {
    // Calculate content hash
    const contentHash = createHash('sha256').update(content).digest();
    
    let cid: string;
    let simulated = false;
    
    try {
      // Try actual IPFS storage
      const ipfsClient = await this.getClient();
      const result = await ipfsClient.add(content);
      cid = result.cid.toString();
    } catch (error) {
      // Fallback to simulated CID if IPFS is not available
      cid = `Qm${contentHash.toString('hex').substring(0, 44)}`;
      simulated = true;
    }
    
    const descriptor = new ContentDescriptor();
    descriptor.uri = `ipfs:${cid}`;
    descriptor.contentType = contentType;
    descriptor.contentLength = BigInt(content.length);
    descriptor.contentHash = contentHash;
    
    descriptor.metadata['storage_provider'] = 'ipfs';
    descriptor.metadata['gateway_url'] = this.gatewayUrl;
    if (simulated) {
      descriptor.metadata['simulated'] = 'true';
    }
    
    return descriptor;
  }

  async retrieve(descriptor: ContentDescriptor): Promise<Uint8Array> {
    if (!this.canHandle(descriptor.uri)) {
      throw new ValidationError(`Invalid IPFS URI: ${descriptor.uri}`);
    }

    // Extract CID from URI
    const cid = descriptor.uri.substring(5); // Remove "ipfs:" prefix
    
    try {
      const ipfsClient = await this.getClient();
      const chunks: Uint8Array[] = [];
      for await (const chunk of ipfsClient.cat(cid)) {
        chunks.push(chunk);
      }
      
      // Combine chunks into single Uint8Array
      const totalLength = chunks.reduce((acc, chunk) => acc + chunk.length, 0);
      const result = new Uint8Array(totalLength);
      let offset = 0;
      for (const chunk of chunks) {
        result.set(chunk, offset);
        offset += chunk.length;
      }
      return result;
    } catch (error: any) {
      throw new Error(`Failed to retrieve content from IPFS: ${error.message}`);
    }
  }

  canHandle(uri: string): boolean {
    return uri.startsWith('ipfs:');
  }
}

/**
 * Redis content storage implementation.
 */
export class RedisStorage extends ContentStorage {
  private redisUrl: string;
  private keyPrefix: string;
  private _client: any;

  constructor(redisUrl: string = 'redis://localhost:6379', keyPrefix: string = 'postfiat:content') {
    super();
    this.redisUrl = redisUrl;
    this.keyPrefix = keyPrefix;
  }

  private async getClient() {
    if (!this._client) {
      try {
        const Redis = await import('ioredis');
        this._client = new Redis.default(this.redisUrl);
      } catch (error) {
        throw new Error(
          'ioredis is required for Redis storage. ' +
          'Install with: npm install ioredis'
        );
      }
    }
    return this._client;
  }

  private contentKey(contentHash: Uint8Array): string {
    return `${this.keyPrefix}:${Buffer.from(contentHash).toString('hex')}`;
  }

  async store(content: Uint8Array, contentType: string): Promise<ContentDescriptor> {
    // Calculate content hash
    const contentHash = createHash('sha256').update(content).digest();
    
    // Store content in Redis
    const key = this.contentKey(contentHash);
    
    // Store content and metadata together
    const contentData = {
      content: Buffer.from(content).toString('base64'),
      content_type: contentType,
      content_length: content.length,
      content_hash: Buffer.from(contentHash).toString('hex')
    };
    
    try {
      const client = await this.getClient();
      await client.set(key, JSON.stringify(contentData));
      
      const descriptor = new ContentDescriptor();
      descriptor.uri = `redis:${Buffer.from(contentHash).toString('hex')}`;
      descriptor.contentType = contentType;
      descriptor.contentLength = BigInt(content.length);
      descriptor.contentHash = contentHash;
      
      descriptor.metadata['storage_provider'] = 'redis';
      descriptor.metadata['redis_key'] = key;
      descriptor.metadata['redis_url'] = this.redisUrl;
      
      return descriptor;
    } catch (error: any) {
      throw new Error(`Failed to store content in Redis: ${error.message}`);
    }
  }

  async retrieve(descriptor: ContentDescriptor): Promise<Uint8Array> {
    if (!this.canHandle(descriptor.uri)) {
      throw new ValidationError(`Invalid Redis URI: ${descriptor.uri}`);
    }

    // Extract content hash from URI
    const contentHashHex = descriptor.uri.substring(6); // Remove "redis:" prefix
    const contentHash = Buffer.from(contentHashHex, 'hex');
    
    const key = this.contentKey(contentHash);
    
    try {
      const client = await this.getClient();
      const contentData = await client.get(key);
      
      if (!contentData) {
        throw new Error(`Content not found in Redis: ${descriptor.uri}`);
      }
      
      const data = JSON.parse(contentData);
      const content = Buffer.from(data.content, 'base64');
      
      // Verify content hash
      const actualHash = createHash('sha256').update(content).digest();
      if (!Buffer.from(actualHash).equals(contentHash)) {
        throw new Error(`Content hash mismatch for ${descriptor.uri}`);
      }
      
      return new Uint8Array(content);
    } catch (error: any) {
      throw new Error(`Failed to retrieve content from Redis: ${error.message}`);
    }
  }

  canHandle(uri: string): boolean {
    return uri.startsWith('redis:');
  }
}

/**
 * Inline content storage - stores content directly in descriptor.
 */
export class InlineStorage extends ContentStorage {
  async store(content: Uint8Array, contentType: string): Promise<ContentDescriptor> {
    // Calculate content hash
    const contentHash = createHash('sha256').update(content).digest();
    
    // Encode content as base64 for inline storage
    const contentB64 = Buffer.from(content).toString('base64');
    
    const descriptor = new ContentDescriptor();
    descriptor.uri = `data:${contentType};base64,${contentB64}`;
    descriptor.contentType = contentType;
    descriptor.contentLength = BigInt(content.length);
    descriptor.contentHash = contentHash;
    
    descriptor.metadata['storage_provider'] = 'inline';
    descriptor.metadata['content_data'] = contentB64;
    
    return descriptor;
  }

  async retrieve(descriptor: ContentDescriptor): Promise<Uint8Array> {
    if (!this.canHandle(descriptor.uri)) {
      throw new ValidationError(`Invalid data URI: ${descriptor.uri}`);
    }

    try {
      // Parse data URI: data:contentType;base64,data
      const match = descriptor.uri.match(/^data:([^;]+);base64,(.+)$/);
      if (!match) {
        throw new Error('Invalid data URI format');
      }
      
      const contentB64 = match[2];
      const content = Buffer.from(contentB64, 'base64');
      
      // Verify content hash
      const actualHash = createHash('sha256').update(content).digest();
      if (!Buffer.from(actualHash).equals(Buffer.from(descriptor.contentHash))) {
        throw new Error('Content hash mismatch for inline content');
      }
      
      return new Uint8Array(content);
    } catch (error: any) {
      throw new Error(`Failed to retrieve inline content: ${error.message}`);
    }
  }

  canHandle(uri: string): boolean {
    return uri.startsWith('data:');
  }
}

/**
 * Storage-agnostic multipart message chunking.
 * 
 * This storage backend has no knowledge of the underlying ledger.
 * It just creates multipart envelopes that can be stored anywhere.
 */
export class MultipartStorage extends ContentStorage {
  private maxPartSize: number;

  constructor(maxPartSize: number = 800) {
    super();
    this.maxPartSize = maxPartSize;
  }

  async store(content: Uint8Array, contentType: string): Promise<ContentDescriptor> {
    // Generate unique message ID
    const messageId = crypto.randomUUID();
    
    // Calculate content hash
    const contentHash = createHash('sha256').update(content).digest();
    
    // Calculate number of parts needed
    const totalParts = Math.ceil(content.length / this.maxPartSize);
    
    // Create multipart URI - describes how to find all parts
    const uri = `multipart:${messageId}`;
    
    const descriptor = new ContentDescriptor();
    descriptor.uri = uri;
    descriptor.contentType = contentType;
    descriptor.contentLength = BigInt(content.length);
    descriptor.contentHash = contentHash;
    
    descriptor.metadata['storage_provider'] = 'multipart';
    descriptor.metadata['message_id'] = messageId;
    descriptor.metadata['total_parts'] = totalParts.toString();
    descriptor.metadata['part_size'] = this.maxPartSize.toString();
    
    return descriptor;
  }

  /**
   * Create multipart envelopes for multipart storage.
   */
  async createPartEnvelopes(
    content: Uint8Array,
    descriptor: ContentDescriptor,
    encryptionMode: EncryptionMode = EncryptionMode.PROTECTED,
    baseMetadata: { [key: string]: string } = {}
  ): Promise<Set<Envelope>> {
    if (!this.canHandle(descriptor.uri)) {
      throw new ValidationError(`Invalid multipart URI: ${descriptor.uri}`);
    }

    const messageId = descriptor.metadata['message_id'];
    const totalParts = parseInt(descriptor.metadata['total_parts'] || '0');

    if (!messageId || totalParts === 0) {
      throw new ValidationError('Missing multipart metadata');
    }

    const envelopes = new Set<Envelope>();

    for (let i = 0; i < content.length; i += this.maxPartSize) {
      const partNumber = Math.floor(i / this.maxPartSize) + 1;
      const chunkContent = content.slice(i, i + this.maxPartSize);

      // Create multipart message part
      const multipart = new MultiPartMessagePart();
      multipart.messageId = messageId;
      multipart.partNumber = partNumber;
      multipart.totalParts = totalParts;
      multipart.content = chunkContent;
      multipart.completeMessageHash = Buffer.from(descriptor.contentHash).toString('base64');

      // Create envelope for this part
      const messageBytes = multipart.toBinary();

      const metadata = { ...baseMetadata };
      metadata['multipart'] = `${partNumber}/${totalParts}`;
      metadata['message_id'] = messageId;

      const envelope = new Envelope();
      envelope.version = 1;
      envelope.contentHash = createHash('sha256').update(messageBytes).digest('hex');
      envelope.messageType = MessageType.MULTIPART_MESSAGE_PART;
      envelope.encryption = encryptionMode;
      envelope.message = messageBytes;
      
      Object.entries(metadata).forEach(([key, value]) => {
        envelope.metadata[key] = value;
      });

      envelopes.add(envelope);
    }

    return envelopes;
  }

  async retrieve(descriptor: ContentDescriptor): Promise<Uint8Array> {
    throw new Error('Multipart retrieval requires collecting all envelope parts by message_id');
  }

  canHandle(uri: string): boolean {
    return uri.startsWith('multipart:');
  }
}

/**
 * HTTP/HTTPS content storage implementation.
 */
export class HTTPStorage extends ContentStorage {
  async store(content: Uint8Array, contentType: string): Promise<ContentDescriptor> {
    // This would typically POST to a content server
    throw new Error('HTTP storage not yet implemented');
  }

  async retrieve(descriptor: ContentDescriptor): Promise<Uint8Array> {
    if (!this.canHandle(descriptor.uri)) {
      throw new ValidationError(`Invalid HTTP URI: ${descriptor.uri}`);
    }

    // TODO: Implement HTTP retrieval
    // const response = await fetch(descriptor.uri);
    // return new Uint8Array(await response.arrayBuffer());
    
    throw new Error('HTTP retrieval not yet implemented');
  }

  canHandle(uri: string): boolean {
    return uri.startsWith('http://') || uri.startsWith('https://');
  }
}

/**
 * Composite storage that delegates to appropriate backend.
 */
export class CompositeStorage extends ContentStorage {
  private storages: ContentStorage[];

  constructor(storages: ContentStorage[]) {
    super();
    this.storages = storages;
  }

  async store(content: Uint8Array, contentType: string): Promise<ContentDescriptor> {
    if (this.storages.length === 0) {
      throw new ValidationError('No storage backends configured');
    }

    // Use first storage backend for storing
    return this.storages[0].store(content, contentType);
  }

  async retrieve(descriptor: ContentDescriptor): Promise<Uint8Array> {
    for (const storage of this.storages) {
      if (storage.canHandle(descriptor.uri)) {
        return storage.retrieve(descriptor);
      }
    }

    throw new ValidationError(`No storage backend can handle URI: ${descriptor.uri}`);
  }

  canHandle(uri: string): boolean {
    return this.storages.some(storage => storage.canHandle(uri));
  }
}